natscontext/context.go

```diff
@@ -0,0 +1,336 @@
+	} else {
+		nctx, err = newFromName(name)
+	}
+	if err != nil {
+		return nil, Settings{}, err
+	}
+
+	nc, err := nctx.connect(opts...)
+	if err != nil {
+		return nil, Settings{}, err
+	}
+
+	return nc, *nctx.config, nil
+}
+
+func (c *context) connect(opts ...nats.Option) (*nats.Conn, error) {
+	nopts, err := c.nATSOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return nats.Connect(c.config.URL, nopts...)
+}
+
+// nATSOptions creates NATS client configuration based on the contents of the context
+func (c *context) nATSOptions(opts ...nats.Option) ([]nats.Option, error) {
+	var nopts []nats.Option
+
+	switch {
+	case c.config.User != "":
+		nopts = append(nopts, nats.UserInfo(c.config.User, c.config.Password))
+	case c.config.Creds != "":
+		nopts = append(nopts, nats.UserCredentials(expandHomedir(c.config.Creds)))
+
+	case c.config.NKey != "":
+		nko, err := nats.NkeyOptionFromSeed(expandHomedir(c.config.NKey))
+		if err != nil {
+			return nil, err
+		}
+
+		nopts = append(nopts, nko)
+	}
+
+	if c.config.Token != "" {
+		nopts = append(nopts, nats.Token(expandHomedir(c.config.Token)))
+	}
+
+	if c.config.Cert != "" && c.config.Key != "" {
+		nopts = append(nopts, nats.ClientCert(expandHomedir(c.config.Cert), expandHomedir(c.config.Key)))
+	}
+
+	if c.config.CA != "" {
+		nopts = append(nopts, nats.RootCAs(expandHomedir(c.config.CA)))
+	}
+
+	if c.config.SocksProxy != "" {
+		nopts = append(nopts, nats.SetCustomDialer(c.SOCKSDialer()))
+	}
+
+	if c.config.InboxPrefix != "" {
+		nopts = append(nopts, nats.CustomInboxPrefix(c.config.InboxPrefix))
+	}
+
+	if c.config.TLSFirst {
+		nopts = append(nopts, nats.TLSHandshakeFirst())
+	}
+
+	if c.config.WinCertStoreType != "" {
+		return nil, fmt.Errorf("windows cert stores are not supported")
+	}
+
+	nopts = append(nopts, opts...)
+
+	return nopts, nil
+}
+
+func expandHomedir(path string) string {
+	if path[0] != '~' {
+		return path
+	}
+
+	usr, err := user.Current()
+	if err != nil {
+		return path
+	}
+
+	return strings.Replace(path, "~", usr.HomeDir, 1)
+}
+
+func newFromName(name string) (*context, error) {
+	c := &context{
+		Name:   name,
+		config: &Settings{},
+	}
+
+	err := c.loadActiveContext()
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func newFromFile(filename string) (*context, error) {
+	c := &context{
+		Name:   strings.TrimSuffix(filepath.Base(filename), filepath.Ext(filename)),
+		config: &Settings{},
+		path:   filename,
+	}
+
+	err := c.loadActiveContext()
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func readCtxFromFile(file string) string {
+	parent, err := parentDir()
+	if err != nil {
+		return ""
+	}
+
+	currentFile := filepath.Join(parent, "nats", file)
+
+	_, err = os.Stat(currentFile)
+	if os.IsNotExist(err) {
+		return ""
+	}
+
+	fc, err := os.ReadFile(currentFile)
+	if err != nil {
+		return ""
+	}
+
+	return strings.TrimSpace(string(fc))
+}
+
+func selectedContext() string {
+	return readCtxFromFile("context.txt")
```

Comment:
I had no idea we had this default, is it documented anywhere? If not, we probably should :)

Reply:
Thanks will get to the other comments - for this, the ADR fully specifies it https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-21.md

Reply:
`nats ctx select` creates this file :)

Reply:
Thanks!